modules/dex/libdexc/adapter.go

```diff
@@ -0,0 +1,264 @@
+		"UpdateWallet": c.updateWallet,
+		"User":         c.user,
+		"Register":     c.register,
+		"Login":        c.login,
+		"Logout":       c.logout,
+		"GetFee":       c.getFee,
+	}
+
+	return c
+}
+
+func (c *CoreAdapter) startCore(raw json.RawMessage) error {
+	if !atomic.CompareAndSwapUint32(&c.inited, 0, 1) {
+		return fmt.Errorf("already initialized")
+	}
+
+	form := new(struct {
+		DBPath   string      `json:"dbPath"`
+		Net      dex.Network `json:"net"`
+		LogLevel slog.Level  `json:"logLevel"`
+	})
+	if err := json.Unmarshal(raw, form); err != nil {
+		return err
+	}
+
+	err := os.MkdirAll(filepath.Dir(form.DBPath), 0777)
+	if err != nil {
+		return err
+	}
+
+	c.ctx, c.kill = context.WithCancel(context.Background())
+	ccore, err := core.New(&core.Config{
+		DBPath: form.DBPath,
+		Net:    form.Net,
+		Logger: dex.StdOutLogger("CORE", form.LogLevel),
+	})
+	if err != nil {
+		return fmt.Errorf("error creating client core: %v", err)
+	}
+	c.core = ccore
+	c.logLevel = form.LogLevel
+
+	go ccore.Run(c.ctx)
+	<-ccore.Ready()
+
+	return nil
+}
+
+func (c *CoreAdapter) startServer(raw json.RawMessage) (string, error) {
+	if !atomic.CompareAndSwapUint32(&c.serverRunning, 0, 1) {
+		return "", fmt.Errorf("already initialized")
+	}
+	form := new(struct {
+		WebAddr string `json:"webaddr"`
+		SiteDir string `json:"sitedir"`
+	})
+	if err := json.Unmarshal(raw, form); err != nil {
+		return "", err
+	}
+	webSrv, err := webserver.New(c.core, form.WebAddr, form.SiteDir, dex.StdOutLogger("SRVR", c.logLevel), false)
+	if err != nil {
+		return "", fmt.Errorf("Error creating web server: %v", err)
+	}
+	cm := dex.NewConnectionMaster(webSrv)
+	err = cm.Connect(c.ctx)
+	if err != nil {
+		return "", fmt.Errorf("Error starting web server: %v", err)
+	}
+	c.webServer = cm
+	go func() {
+		defer atomic.StoreUint32(&c.serverRunning, 0)
+		cm.Wait()
+	}()
+	return "", nil
+}
+
+func (c *CoreAdapter) shutdown(json.RawMessage) (string, error) {
+	if !atomic.CompareAndSwapUint32(&c.inited, 1, 0) || c.kill == nil {
+		return "", fmt.Errorf("already shut down")
+	}
+	if !atomic.CompareAndSwapUint32(&c.serverRunning, 1, 0) {
+		return "", fmt.Errorf("server not running initialized")
+	}
+	c.kill()
+	return "", nil
+}
+
+func (c *CoreAdapter) handlers(funcName string) (callHandler, callHandler) {
+	return c.preInitMethods[funcName], c.directMethods[funcName]
+}
+
+func (c *CoreAdapter) run(callData *CallData) (string, error) {
+	switch preInitHandler, coreHandler := c.handlers(callData.Function); {
+	case callData.Function == "startCore":
+		return "", c.startCore(callData.Params)
+	case preInitHandler != nil:
+		return preInitHandler(callData.Params)
+	case atomic.LoadUint32(&c.inited) == 0:
+		return "", fmt.Errorf("not initialized")
+	case c.core == nil:
+		return "", fmt.Errorf("core not constructed. probably an initialization error")
+	case coreHandler != nil:
+		return coreHandler(callData.Params)
+	}
+	return "", fmt.Errorf("no method %q", callData.Function)
```

Comment:
I guess `callData` is guaranteed to not be nil?

Reply:
I'll put in a basic not nil check at the top